lib/kube/proxy/transport.go

```diff
@@ -0,0 +1,419 @@
+	"github.com/gravitational/teleport/api/types"
+	"github.com/gravitational/teleport/lib/auth"
+	"github.com/gravitational/teleport/lib/defaults"
+	"github.com/gravitational/teleport/lib/reversetunnel"
+	"github.com/gravitational/teleport/lib/utils"
+)
+
+// transportForRequest determines the transport to use for a request to a specific
+// Kubernetes cluster. If the cluster is remote or if the servers don't support
+// impersonation, a single transport is used per request. Otherwise, a new
+// transport is used for all requests in order to improve performance.
+// TODO(tigrato): Remove this once all servers support impersonation.
+func (f *Forwarder) transportForRequest(sess *clusterSession) (http.RoundTripper, error) {
+	// If all servers support impersonation, use a new transport for each
+	// request. This will ensure that the client certificate is valid for the
+	// server that the request is being sent to.
+	if !sess.teleportCluster.isRemote && allServersSupportImpersonation(sess.kubeServers) {
+		return f.transportForRequestWithImpersonation(sess)
+	}
+	// Otherwise, use a single transport per request.
+	return f.transportForRequestWithoutImpersonation(sess)
+}
+
+// DialContextFunc is a context network dialer function that returns a network connection
+type DialContextFunc func(context.Context, string, string) (net.Conn, error)
+
+// transportForRequestWithoutImpersonation returns a transport that does not
+// support impersonation. This is used when the cluster is remote or if the
+// servers don't support impersonation in order to ensure that the client request
+// can be routed to the correctly.
+//
+// DELETE in 14.0.0
+// TODO(tigrato): Remove this once all servers support impersonation.
+func (f *Forwarder) transportForRequestWithoutImpersonation(sess *clusterSession) (http.RoundTripper, error) {
+	if sess.creds != nil {
+		return sess.creds.getTransport(sess.upgradeToHTTP2), nil
+	}
+	transport := newTransport(sess.DialWithContext, sess.tlsConfig)
+	if !sess.upgradeToHTTP2 {
+		return tracehttp.NewTransport(transport), nil
+	}
+	if err := http2.ConfigureTransport(transport); err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return tracehttp.NewTransport(transport), nil
+}
+
+// transportForRequestWithImpersonation returns a transport that supports
+// impersonation. This is used when the cluster is local and all servers support
+// impersonation. This allows the client to reuse the same transport for all
+// requests to the cluster in order to improve performance.
+// The transport is cached in the forwarder so that it can be reused for future
+// requests. If the transport is not cached, a new one is created and cached.
+// The transport is cached using the cluster name as the key.
+func (f *Forwarder) transportForRequestWithImpersonation(sess *clusterSession) (http.RoundTripper, error) {
+	// If the cluster is remote, the key is the teleport cluster name.
```

Comment:
It was a typo.

Reply:
Hmm, but isn't it still the case that this is never run if the cluster is remote?

https://github.com/gravitational/teleport/pull/23978#discussion_r1156395259

Reply:
The code is ready for all cases and I will fix the host ca sync using another PR since the change is unrelated to the kube code.
Once the host ca sync lands, I will enable this code path.